File: src/scanner.ts
import { Lox } from "./lox";
import { Token } from "./token";
import { TokenType } from "./types";

export class Scanner {
    private static readonly keywords = new Map([
        ["and", TokenType.AND],
        ["class", TokenType.CLASS],
        ["else", TokenType.ELSE],
        ["false", TokenType.FALSE],
        ["for", TokenType.FOR],
        ["fun", TokenType.FUN],
        ["if", TokenType.IF],
        ["nil", TokenType.NIL],
        ["or", TokenType.OR],
        ["print", TokenType.PRINT],
        ["return", TokenType.RETURN],
        ["super", TokenType.SUPER],
        ["this", TokenType.THIS],
        ["true", TokenType.TRUE],
        ["var", TokenType.VAR],
        ["while", TokenType.WHILE],
    ]);

    private readonly tokens = [] as Token[];

    private start = 0;
    private current = 0;
    private line = 0;
    private col = 1;

    public constructor(public readonly source: string) {}

    public scanTokens() {
        while (!this.isAtEnd()) {
            this.start = this.current;
            this.scanToken();
        }

        this.tokens.push(new Token(TokenType.EOF, "", null, this.line, 1));

        return this.tokens;
    }

    private scanToken() {
        const c = this.advance();
        switch (c) {
            case "(":
                this.addToken(TokenType.LEFT_PAREN);
                break;
            case ")":
                this.addToken(TokenType.RIGHT_PAREN);
                break;
            case "{":
                this.addToken(TokenType.LEFT_BRACE);
                break;
            case "}":
                this.addToken(TokenType.RIGHT_BRACE);
                break;
            case ",":
                this.addToken(TokenType.COMMA);
                break;
            case ".":
                this.addToken(TokenType.DOT);
                break;
            case "-":
                this.addToken(TokenType.MINUS);
                break;
            case "+":
                this.addToken(TokenType.PLUS);
                break;
            case ";":
                this.addToken(TokenType.SEMICOLON);
                break;
            case "*":
                this.addToken(TokenType.STAR);
                break;
            case "!":
                this.addToken(this.match("=") ? TokenType.BANG_EQUAL : TokenType.BANG);
                break;
            case "?":
                this.addToken(TokenType.QUESTION);
                break;
            case ":":
                this.addToken(TokenType.COLON);
                break;
            case "=":
                this.addToken(this.match("=") ? TokenType.EQUAL_EQUAL : TokenType.EQUAL);
                break;
            case "<":
                this.addToken(this.match("=") ? TokenType.LESS_EQUAL : TokenType.LESS);
                break;
            case ">":
                this.addToken(this.match("=") ? TokenType.GREATER_EQUAL : TokenType.GREATER);
                break;
            case "/":
                if (this.match("/")) {
                    while (this.peek() != "\n" && !this.isAtEnd()) this.advance();
                } else if (this.match("*")) {
                    this.comments();
                } else {
                    this.addToken(TokenType.SLASH);
                }
                break;
            case '"':
                this.string();
                break;
            case " ":
            case "\r":
            case "\t":
                break;
            case "\n":
                this.line++;
                this.col = 1;
                break;
            default:
                if (this.isDigit(c)) {
                    this.number();
                } else if (this.isAlpha(c)) {
                    this.identifier();
                } else {
                    Lox.error(this.line, this.col, "Unexpected character.");
                }
                break;
        }
    }

    private string() {
        while (this.peek() != '"' && !this.isAtEnd()) {
            if (this.peek() == "\n") this.line++;

            this.advance();
        }

        if (this.isAtEnd()) return Lox.error(this.line, this.col, "Unterminated string.");

        this.advance();

        const value = this.source.substring(this.start + 1, this.current - 1);

        this.addToken(TokenType.STRING, value);
    }

    private number() {
        while (this.isDigit(this.peek())) this.advance();

        if (this.peek() === "." && this.isDigit(this.peekNext())) {
            this.advance();

            while (this.isDigit(this.peek())) this.advance();
        }

        this.addToken(TokenType.NUMBER, Number(this.source.substring(this.start, this.current)));
    }

    private identifier() {
        while (this.isAlphaNumeric(this.peek())) this.advance();

        const text = this.source.substring(this.start, this.current);

        const type = Scanner.keywords.get(text) ?? TokenType.IDENTIFIER;

        this.addToken(type);
    }

    private comments() {
        while (!this.isAtEnd() && !(this.peek() === "*" && this.peekNext() == "/")) {
            if (this.peek() === "\n") this.line++;
            this.advance();
        }

        if (this.isAtEnd()) {
            Lox.error(this.line, this.col, "Unterminated comment.");
            return;
        }

        this.advance();
        if (!this.isAtEnd()) this.advance();
    }

    private peek() {
        if (this.isAtEnd()) return "\0";
        return this.source[this.current];
    }

    private peekNext() {
        if (this.current + 1 >= this.source.length) return "\0";

        return this.source.charAt(this.current + 1);
    }

    private advance() {
        return this.source[this.current++];
    }

    private match(expected: string) {
        if (this.isAtEnd()) return false;
        if (this.source.charAt(this.current) != expected) return false;

        this.current++;
        return true;
    }

    private isDigit(c: string) {
        return /^[0-9]$/.test(c);
    }

    private isAlpha(c: string) {
        return /^[a-zA-Z_]$/.test(c);
    }

    private isAlphaNumeric(c: string) {
        return this.isAlpha(c) || this.isDigit(c);
    }

    private addToken(type: TokenType, literal?: unknown) {
        const text = this.source.substring(this.start, this.current);

        this.tokens.push(new Token(type, text, literal ?? null, this.line, this.col));

        this.col += text.length;
    }

    private isAtEnd() {
        return this.current >= this.source.length;
    }
}

File: rust/src/scanner.rs
#[derive(Debug, PartialEq, Eq, Hash, Clone, Copy)]
pub enum TokenKind {
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    Comma,
    Dot,
    Minus,
    Plus,
    Semicolon,
    Slash,
    Star,
    Bang,
    BangEqual,
    Equal,
    EqualEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,
    Identifier,
    String,
    Number,
    And,
    Class,
    Else,
    False,
    Fun,
    For,
    If,
    Nil,
    Or,
    Print,
    Return,
    Super,
    This,
    True,
    Var,
    While,
    Err,
    Eof,
}

#[derive(Clone, Copy)]
pub struct Token {
    pub kind: TokenKind,
    pub start: usize,
    pub length: usize,
    pub line: usize,
    pub col: usize,
}

pub trait TokenScanner {
    fn next_token(&mut self) -> Token;
    fn is_at_end(&mut self) -> bool;
    fn make_token(&mut self, kind: TokenKind) -> Token;
    fn error_token(&mut self, msg: String) -> Token;
    fn advance(&mut self) -> char;
    fn match_char(&mut self, expected: char) -> bool;
    fn skip_whitespace(&mut self);
    fn peek(&mut self) -> char;
    fn peek_next(&mut self) -> char;
    fn is_digit(&mut self, expected: char) -> bool;
    fn is_alpha(&mut self, expected: char) -> bool;
    fn string(&mut self) -> Token;
    fn number(&mut self) -> Token;
    fn identifier(&mut self) -> Token;
}

pub struct Scanner {
    source: String,
    start: usize,
    current: usize,
    line: usize,
    col: usize,
}

impl TokenScanner for Scanner {
    fn next_token(&mut self) -> Token {
        self.skip_whitespace();

        self.start = self.current;

        if self.is_at_end() {
            return self.make_token(TokenKind::Eof);
        }

        let c = self.advance();

        if self.is_digit(c) {
            return self.number();
        }

        if self.is_alpha(c) {
            return self.identifier();
        }

        match c {
            '(' => self.make_token(TokenKind::LeftParen),
            ')' => self.make_token(TokenKind::RightParen),
            '{' => self.make_token(TokenKind::LeftBrace),
            '}' => self.make_token(TokenKind::RightBrace),
            ',' => self.make_token(TokenKind::Comma),
            '.' => self.make_token(TokenKind::Dot),
            '-' => self.make_token(TokenKind::Minus),
            '+' => self.make_token(TokenKind::Plus),
            ';' => self.make_token(TokenKind::Semicolon),
            '/' => self.make_token(TokenKind::Slash),
            '*' => self.make_token(TokenKind::Star),
            '!' => {
                if self.match_char('=') {
                    self.make_token(TokenKind::BangEqual)
                } else {
                    self.make_token(TokenKind::Bang)
                }
            }
            '=' => {
                if self.match_char('=') {
                    self.make_token(TokenKind::EqualEqual)
                } else {
                    self.make_token(TokenKind::Equal)
                }
            }
            '>' => {
                if self.match_char('=') {
                    self.make_token(TokenKind::GreaterEqual)
                } else {
                    self.make_token(TokenKind::Greater)
                }
            }
            '<' => {
                if self.match_char('=') {
                    self.make_token(TokenKind::LessEqual)
                } else {
                    self.make_token(TokenKind::Less)
                }
            }
            '"' => self.string(),
            _ => self.error_token("Unexpected character.".to_string()),
        }
    }

    fn string(&mut self) -> Token {
        while self.peek() != '"' && !self.is_at_end() {
            self.advance();
        }

        if self.is_at_end() {
            return self.error_token("Unterminated string.".to_string());
        }

        self.advance();

        self.make_token(TokenKind::String)
    }

    fn number(&mut self) -> Token {
        let mut c = self.peek();

        while self.is_digit(c) {
            self.advance();

            c = self.peek();
        }

        c = self.peek_next();

        if self.peek() == '.' && self.is_digit(c) {
            self.advance();

            c = self.peek();

            while self.is_digit(c) {
                self.advance();

                c = self.peek();
            }
        }

        self.make_token(TokenKind::Number)
    }

    fn identifier(&mut self) -> Token {
        let mut c = self.peek();

        while self.is_alpha(c) || self.is_digit(c) {
            self.advance();

            c = self.peek();
        }

        let text = self.source[self.start..self.current].to_string();

        match text.as_ref() {
            "and" => self.make_token(TokenKind::And),
            "class" => self.make_token(TokenKind::Class),
            "else" => self.make_token(TokenKind::Else),
            "false" => self.make_token(TokenKind::False),
            "fun" => self.make_token(TokenKind::Fun),
            "for" => self.make_token(TokenKind::For),
            "if" => self.make_token(TokenKind::If),
            "nil" => self.make_token(TokenKind::Nil),
            "or" => self.make_token(TokenKind::Or),
            "print" => self.make_token(TokenKind::Print),
            "return" => self.make_token(TokenKind::Return),
            "super" => self.make_token(TokenKind::Super),
            "this" => self.make_token(TokenKind::This),
            "true" => self.make_token(TokenKind::True),
            "var" => self.make_token(TokenKind::Var),
            "while" => self.make_token(TokenKind::While),
            _ => self.make_token(TokenKind::Identifier),
        }
    }

    fn is_at_end(&mut self) -> bool {
        return self.current >= self.source.len();
    }

    fn make_token(&mut self, kind: TokenKind) -> Token {
        Token {
            kind,
            start: self.start,
            length: self.current - self.start,
            line: self.line,
            col: self.col,
        }
    }

    fn error_token(&mut self, msg: String) -> Token {
        Token {
            kind: TokenKind::Err,
            start: self.start,
            length: msg.len(),
            line: self.line,
            col: self.col,
        }
    }

    fn advance(&mut self) -> char {
        self.current += 1;
        self.col += 1;

        let c = self.source.chars().nth(self.current - 1).unwrap();

        if c == '\n' {
            self.line += 1;
            self.col = 1;
        }

        return c;
    }

    fn match_char(&mut self, expected: char) -> bool {
        if self.is_at_end() {
            return false;
        }

        if self.peek() != expected {
            return false;
        }

        self.current += 1;

        return true;
    }

    fn skip_whitespace(&mut self) {
        while !self.is_at_end() {
            let c = self.peek();

            if c == ' ' || c == '\r' || c == '\t' || c == '\n' {
                self.advance();
            } else if c == '/' && self.peek_next() == '/' {
                while self.peek() != '\n' && !self.is_at_end() {
                    self.advance();
                }
            } else {
                break;
            }
        }
    }

    fn peek(&mut self) -> char {
        if self.current >= self.source.len() {
            return '\0';
        }

        return self.source.chars().nth(self.current).unwrap();
    }

    fn peek_next(&mut self) -> char {
        if self.current + 1 >= self.source.len() {
            return '\0';
        }

        return self.source.chars().nth(self.current + 1).unwrap();
    }

    fn is_digit(&mut self, expected: char) -> bool {
        expected >= '0' && expected <= '9'
    }

    fn is_alpha(&mut self, expected: char) -> bool {
        (expected >= 'a' && expected <= 'z')
            || (expected >= 'A' && expected <= 'Z')
            || expected == '_'
    }
}

pub fn init_scanner(source: String) -> Scanner {
    Scanner {
        source,
        start: 0,
        current: 0,
        line: 1,
        col: 1,
    }
}
